### ISSUE TITLE
Firehose-to-syslog crashing due to websocket: close 1000 (normal)

### ISSUE THREAD

**Issue by j-kaplan** [cloudfoundry-community/firehose-to-syslog#159]
We're running firehose-to-syslog v3.3.0 and having issues with the nozzles constantly crashing due to websocket: close 1000 (normal) errors. 

In our largest PCF deployment we're currently running PCF 1.11.13 with 24 Doppler servers, 12 Trafficcontrollers and 12 f2s nozzle instances.

We're currently averaging around 5k logs / second with peaks up to 8k/second.

I've attached a dump of recent logs with the system domain redacted.

[firehose-to-syslog-3.3.0.log](https://github.com/cloudfoundry-community/firehose-to-syslog/files/1338844/firehose-to-syslog-3.3.0.log)

**Comment by shinji62** [comment]
@enocom Please I would like why the Firehose is disconnecting nozzle.

I think if the nozzle could not keep up firehose will just the drop the logs with a warning message, but deconnexion is not the best ...

**Comment by enocom** [comment]
Hi @shinji62.

One of the core principles of Loggregator's design is that it will never push back on upstream writers. If for instance a downstream reader is reading slowly and Loggregator simply waited for the reader to finish, that slowdown would eventually propagate back upstream and might even slow down a running application or even Diego itself. As a result, Loggregator does its best to provide fast writes to upstream writers.

One consequence of this design is that Loggregator aggressively disconnects slow readers. We have seen this happen on slow firehose nozzles repeatedly. 

What can we do about it?

Recent documentation for Loggregator makes some suggestions. See [here](https://docs.pivotal.io/pivotalcf/1-12/loggregator/log-ops-guide.html). I'll summarize some key points.

First, we can ensure the nozzle is properly scaled to match Loggregator's size. In short, the number of nozzle instances should match the number of traffic controller instances.

Second, we can ensure that in the case of this nozzle, the configured syslog drain is as fast as possible. The details of how to do that naturally depend on a particular drain.

Finally, there are some code changes we might make in the nozzle here.

First, you should buffer messages from the firehose. In particular, [this](https://github.com/cloudfoundry-community/firehose-to-syslog/blob/8f42b08e9bdf531e15f5ed7ba2c42256892638fd/firehoseclient/firehoseclient.go#L66-L67) write into a channel may block if the consumer of the channel is not going fast enough. We have used [this ring buffer library](https://github.com/cloudfoundry/diodes) to buffer reads.

Second, the event routing locks around network IO [here](https://github.com/cloudfoundry-community/firehose-to-syslog/blob/8f42b08e9bdf531e15f5ed7ba2c42256892638fd/eventRouting/eventrouting_default.go#L70-L79). Locking around IO is super expensive and will prevent Go from running as fast as it could. Removing the locking around IO will require some design changes, but may also provide some performance gains.

I hope all this helps. We are aware the firehose API leaves something to be desired. Currently, we are in the early stages of building out a v2 API with better support for nozzles. Keep on an eye on our backlog or even reach out on [Slack](https://cloudfoundry.slack.com) if you want to chat.

**Comment by shinji62** [comment]
Thanks @enocom I start implementing the diodes thingy https://github.com/cloudfoundry-community/firehose-to-syslog/tree/feature/adding_ring_buffer

**Comment by enocom** [comment]
@shinji62 Beware the size of the diode. The maximum size of a log messages is 64 KB. If you fill up your diode with 100,000 logs, your nozzle will potentially need ~6.4 GB of memory just for the diode: 64 KB * 100,000 = 6,400,000,000 bytes = 6.4GB). We recommend much smaller sizes for the diodes.

Also, it would be good to have some tests around these important code paths.

**Comment by j-kaplan** [comment]
@shinji62 I see you have made some progress in implementing the buffering. Is that in a state where it would be ready for testing? We would be happy to help test at any point since this is an ongoing problem for us. Thanks!

**Comment by enocom** [comment]
@shinji62 @j-kaplan It's important to understand that even with buffering, if the load on the nozzle is greater than the downstream writers can handle. the buffer will drop messages to prevent a disconnect. Also, removing the locking around IO that I mentioned is just as important.

**Comment by shinji62** [comment]
@enocom Yes we undertand is important, unlike you I am not dedicated to the logs part of CF. So I will remove the mutex around stats soon.

To be honest I need to rewrite some big part before implementing somethings better. But time is missing ..


@j-kaplan I have been testing this buffer in PROD since 1 week and no crash so far. You will have to define the buffer to your need and to your available memory. So you can try to compile the buffer branch and use it. Default buffer is quite big, better to reduce it.

**Comment by shinji62** [comment]
@j-kaplan I remove the mutex for the routing event, there is a new stats endpoint `curl  http://localhost:8080/stats/app` There is a data-race in the stats, which I will fix later. But this do not hurt logs processing and metrics .

**Comment by j-kaplan** [comment]
@shinji62 I have started testing the nozzle in one environment and it is a definite improvement with the buffering. We are seeing large amounts of dropped logs from the buffers, but with the buffering its clear that is caused by issues on the syslog side which is currently under investigation. Even with the buffers dropping logs, we still have increased reliability with these fixes since the nozzle crashing has decreased significantly and the instances don't have to constantly refill their caches on start with the large number of apps in this environment (~1800 apps).

I have noticed a few things after observing it for the day.

We occasionally receive this panic about 5-6 minutes after getting a firehose exception and seeing the `Stopping Channel` log message.
```
2017-10-12T15:13:39.57-0600 [APP/PROC/WEB/8] ERR [2017-10-12 21:13:39.573905846 +0000 UTC] Exception occurred! Message: Missed Logs  Details: 10000
2017-10-12T15:13:40.23-0600 [APP/PROC/WEB/14] ERR [2017-10-12 21:13:40.239238738 +0000 UTC] Exception occurred! Message: Error while reading from the firehose Details: websocket: close 1008 (policy violation): Client did not respond to ping before keep-alive timeout expired.
2017-10-12T15:13:40.23-0600 [APP/PROC/WEB/14] ERR [2017-10-12 21:13:40.239305047 +0000 UTC] Exception occurred! Message: Failed connecting to Firehose...Please check settings and try again! Details: 
2017-10-12T15:13:40.23-0600 [APP/PROC/WEB/14] OUT [2017-10-12 21:13:40.239292826 +0000 UTC] Stopping Channel 
2017-10-12T15:13:40.23-0600 [APP/PROC/WEB/14] ERR [2017-10-12 21:13:40.239281543 +0000 UTC] Exception occurred! Message: Closing connection with traffic controller due to error Details: websocket: close 1008 (policy violation): Client did not respond to ping before keep-alive timeout expired.
<snip>
2017-10-12T15:19:37.28-0600 [APP/PROC/WEB/14] OUT Exit status 2
2017-10-12T15:19:37.28-0600 [CELL/14] OUT Exit status 0
2017-10-12T15:19:37.36-0600 [API/2] OUT Process has crashed with type: "web"
2017-10-12T15:19:37.25-0600 [APP/PROC/WEB/14] ERR   /tmp/gobuildpack.gopath733301023/.go/src/github.com/cloudfoundry-community/firehose-to-syslog/main.go:176 +0x114c
2017-10-12T15:19:37.25-0600 [APP/PROC/WEB/14] ERR main.main()
2017-10-12T15:19:37.25-0600 [APP/PROC/WEB/14] OUT [2017-10-12 21:19:37.253954264 +0000 UTC] Stopping Channel 
2017-10-12T15:19:37.25-0600 [APP/PROC/WEB/14] ERR panic: close of closed channel
2017-10-12T15:19:37.25-0600 [APP/PROC/WEB/14] ERR goroutine 1 [running]:
2017-10-12T15:19:37.25-0600 [APP/PROC/WEB/14] ERR github.com/cloudfoundry-community/firehose-to-syslog/firehoseclient.(*FirehoseNozzle).Stop(0xc4203b6840)
2017-10-12T15:19:37.25-0600 [APP/PROC/WEB/14] ERR   /tmp/gobuildpack.gopath733301023/.go/src/github.com/cloudfoundry-community/firehose-to-syslog/firehoseclient/firehoseclient.go:63 +0x4f
2017-10-12T15:19:37.37-0600 [API/2] OUT App instance exited with guid 9f6b6372-8f94-41d2-a546-6b35fcc6d9a5 payload: {"instance"=>"0bf6b646-8da5-41d1-70ef-4903", "index"=>14, "reason"=>"CRASHED", "exit_description"=>"APP/PROC/WEB: Exited with status 2", "crash_count"=>1, "crash_timestamp"=>1507843177298791225, "version"=>"e6f16505-40d3-4beb-ab32-e7ce3560442e"}
```

Also, I was looking into the new stats API you added and I'm not sure if its intended, but I believe the `subinput_buffer` metric becomes inaccurate and keeps increasing if the ring buffer is dropping logs since you don't decrement that counter when the ring-buffer overlaps itself.  

Example output with `LOGS_BUFFER_SIZE=10000`:
```
{"consume":0,"consume_per_sec":0,"consume_fail":0,"consume_http_start_stop":0,"consume_value_metric":0,"consume_counter_event":0,"consume_log_message":67222,"consume_error":0,"consume_container_metric":0,"consume_unknown":0,"ignored":0,"forwarded":0,"publish":67221,"publish_per_sec":155,"publish_fail":0,"slow_consumer_alert":0,"subinupt_buffer":1001290,"instance_id":12}
``` 
Also you have a slight typo in the json representation of that metric:
https://github.com/cloudfoundry-community/firehose-to-syslog/blob/feature/adding_ring_buffer/stats/stats.go#L69


Thanks!

**Comment by shinji62** [comment]
@j-kaplan Thanks for the feedback,I will fix the closing panic. 
For the stats I would like to keep as it is. and I will add a % of buffering is > 100% will be easy to see that the buffer is dropping old logs.

For the caching loading I am aware of this issue and I am thinking about adding a shared storage (redis) I have env. with more than 3000AI and you can image how longs it takes and the stress on the CC..

**Comment by shinji62** [comment]
@j-kaplan I made some changes so please compile the source on the branch and try it.

Thanks

**Comment by j-kaplan** [comment]
Thanks @shinji62. The new changes seem much better. I haven't seen a single crash of the app since pulling the latest, even after receiving the `websocket: close 1008` errors from the firehose.

**Comment by shinji62** [comment]
Good news I will cut a new release soon.

**Comment by shinji62** [comment]
Fixed in 4.0.0 release
